From 4c491daed28a03e6fbad7d9513fa3430b8c19d68 Mon Sep 17 00:00:00 2001
From: Codrin Ciubotariu <codrin.ciubotariu@microchip.com>
Date: Tue, 19 Apr 2022 15:26:05 +0300
Subject: [PATCH 062/471] ASoC: atmel: mchp-i2s-mcc: fix period size
 calculation

The size of the period is used to compute the maxburst. However,
snd_pcm_lib_period_bytes() is 0 because the runtime HW parameters are
computed after the hw_params() callbacks are used. Because of this, to
compute the period size we use params_*() functions.

Fixes: 2c1ab8c6360d ("ASoC: atmel: mchp-i2s-mcc: calculate maxburst based on period size")
Signed-off-by: Codrin Ciubotariu <codrin.ciubotariu@microchip.com>
---
 sound/soc/atmel/mchp-i2s-mcc.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/sound/soc/atmel/mchp-i2s-mcc.c b/sound/soc/atmel/mchp-i2s-mcc.c
index fc127017c523..03bf1d1f4b81 100644
--- a/sound/soc/atmel/mchp-i2s-mcc.c
+++ b/sound/soc/atmel/mchp-i2s-mcc.c
@@ -522,7 +522,8 @@ static int mchp_i2s_mcc_hw_params(struct snd_pcm_substream *substream,
 	unsigned long rate = 0;
 	struct mchp_i2s_mcc_dev *dev = snd_soc_dai_get_drvdata(dai);
 	int sample_bytes = params_physical_width(params) / 8;
-	int period_size = snd_pcm_lib_period_bytes(substream);
+	int period_bytes = params_period_size(params) *
+		params_channels(params) * sample_bytes;
 	int maxburst;
 	u32 mra = 0;
 	u32 mrb = 0;
@@ -533,9 +534,9 @@ static int mchp_i2s_mcc_hw_params(struct snd_pcm_substream *substream,
 	int ret;
 	bool is_playback = (substream->stream == SNDRV_PCM_STREAM_PLAYBACK);
 
-	dev_dbg(dev->dev, "%s() rate=%u format=%#x width=%u channels=%u\n",
+	dev_dbg(dev->dev, "%s() rate=%u format=%#x width=%u channels=%u period_bytes=%d\n",
 		__func__, params_rate(params), params_format(params),
-		params_width(params), params_channels(params));
+		params_width(params), params_channels(params), period_bytes);
 
 	switch (dev->fmt & SND_SOC_DAIFMT_FORMAT_MASK) {
 	case SND_SOC_DAIFMT_I2S:
@@ -644,7 +645,7 @@ static int mchp_i2s_mcc_hw_params(struct snd_pcm_substream *substream,
 	 * We must have the same burst size configured
 	 * in the DMA transfer and in out IP
 	 */
-	maxburst = mchp_i2s_mcc_period_to_maxburst(period_size, sample_bytes);
+	maxburst = mchp_i2s_mcc_period_to_maxburst(period_bytes, sample_bytes);
 	mrb |= MCHP_I2SMCC_MRB_DMACHUNK(maxburst);
 	if (is_playback)
 		dev->playback.maxburst = maxburst;
-- 
2.34.1

